src/Shared/CoreCLRAssemblyLoader.cs

```diff
@@ -51,5 +104,98 @@ public Assembly LoadFromPath(string fullPath)
                 return assembly;
             }
         }
+
+        private Assembly TryGetWellKnownAssembly(AssemblyLoadContext context, AssemblyName assemblyName)
+        {
+            if (!MSBuildLoadContext.WellKnownAssemblyNames.Contains(assemblyName.Name))
+            {
+                return null;
+            }
+
+            // Ensure we are attempting to load a matching version
+            // of the Microsoft.Build.* assembly.
+            assemblyName.Version = _currentAssemblyVersion;
+
+            var searchPaths = new[] { Assembly.GetExecutingAssembly().Location };
+            return TryResolveAssemblyFromPaths(context, assemblyName, searchPaths);
+        }
+
+        private Assembly TryResolveAssembly(AssemblyLoadContext context, AssemblyName assemblyName)
+        {
+            lock (_guard)
+            {
+                Assembly assembly = TryGetWellKnownAssembly(context, assemblyName);
+
+                if (assembly != null)
+                {
+                    return assembly;
+                }
+
+                if (_namesToAssemblies.TryGetValue(assemblyName.FullName, out assembly))
+                {
+                    return assembly;
```

Comment:
Is there no need to check version here?

Reply:
Undeleting, so don't want to change behavior. But no, because .NET Core allows only one assembly per simple name in the same ALC, so if it's already loaded at _some_ version, that's the one you get.